bin/pytorch_inference/CCommandParser.cc

```diff
@@ -152,7 +165,7 @@ void CCommandParser::jsonToRequest(const rapidjson::Document& doc) {
     TUint64Vec arg;
     while (doc.HasMember(varArgName)) {
         const rapidjson::Value& v = doc[varArgName];
-        for (auto itr = v.Begin(); itr != v.End(); ++itr) {
+        for (const auto* itr = v.Begin(); itr != v.End(); ++itr) {
```

Comment:
This looks strange.  It's idiomatic to use iterators by value in for loops.  I think what you've changed it to is working because a RapidJSON array iterator happens to be a raw pointer, but it's a breach of encapsulation to code for that.  Explicitly stating that the `auto` has to be a pointer isn't making it more efficient.  The type `auto` referred to would have been a pointer anyway if that's what a RapidJSON array iterator is.

Reply:
The IDE suggested this change based on the clang tidy rule `clang-tidy(readability-qualified-auto)` I was surprised it compiled TBH then forgot to revert it. 

@valeriy42 do you want to re-evaluate the use of [clang-tidy(readability-qualified-auto)](https://clang.llvm.org/extra/clang-tidy/checks/readability-qualified-auto.html) given the above. Does it align with the repo's coding standards?

Reply:
Digging into the link, the warning comes from [LLVM coding standards](https://llvm.org/docs/CodingStandards.html#beware-unnecessary-copies-with-auto) where they suggest avoiding unnecessary copies and making pointer copies explicit (for readability). 

![image](https://user-images.githubusercontent.com/1292899/111980348-75f53180-8b06-11eb-99a6-5720770177c7.png)

In this trade-off between encapsulation and easiness to uncover unnecessary copy by value, LLVM advice towards latter. AFAICS, in our repo's coding standards this topic is not mentioned explicitly.

Reply:
The standard library classes presumably do something to avoid such encapsulation breaches.  For example, an `std::vector<int>::iterator` _could_ be an `int*`.  It's up to the library implementor whether it is or whether they use a class to implement vector iterators.

So for standard libraries where it _is_ a simple `int*`, using explicit types we could write:

```
    std::vector<int> myVec{1, 2, 3};
    for (int* iter = myVec.begin(); iter != myVec.end(); ++iter) {
        // blah
    }
```

I don't think anyone would say that was best practice.  The equivalent code with explicit types but using the typedefs instead of breaching encapsulation would be:

```
    std::vector<int> myVec{1, 2, 3};
    for (std::vector<int>::iterator iter = myVec.begin(); iter != myVec.end(); ++iter) {
        // blah
    }
```

That is exactly as efficient as using `int*` as the type in the case where `std::vector<int>::iterator` is a simple `int*`.

So if `clang-tidy` isn't recommending replacement of `auto` with `auto*` for some standard container iterators that happen to be pointers that makes me think there is some way that it knows that for iterators encapsulation is more important than revealing what is a pointer.  It makes me think that either this knowledge is hardcoded in `clang-tidy` for certain standard library classes, or else there's something RapidJSON is doing that bypasses some logic in `clang-tidy`.  Possibly it's because the RapidJSON methods for getting iterators are `Begin()` and `End()` instead of `begin()` and `end()`.

Reply:
The second line causes a compilation failure:

```c++
    static_assert(std::is_pointer<rapidjson::GenericValue<rapidjson::UTF8<char>>::ConstValueIterator>::value, 
        "RapidJSON iterator is pointer");

    static_assert(std::is_pointer<std::vector<int>::iterator>::value, 
        "Int iterator is pointer");   
```

The RapidJSON iterator is defined as a pointer in document.h

`typedef const GenericValue* ConstValueIterator; //!< Constant value iterator for iterating in array.`

As @valeriy42 said the rule isn't explicitly defined in `.clang-tidy` it comes as part of the LLVM rule set. For the sake of a few squiggly lines in VSCode there is not point starting to pick those rules apart. 

I've reverted the iterator back to `auto itr` and I've learned something today so we will call that a win and merge as is.